databuilder/query_language.py

```diff
@@ -154,6 +154,14 @@ def filter(self, *args: str, **kwargs: Any) -> BaseTable:  # noqa: A003
             # convert a between filter into its two components
             return self.filter(*args, on_or_after=value[0], on_or_before=value[1])
 
+        if operator in ("equals", "not_equals") and isinstance(
+            value, (Codelist, Column)
+        ):
+            raise TypeError(
+                f"You can only use '{operator}' to filter a column by a single value.\n"
+                f"To filter using a {value.__class__.__name__}, use 'is_in/not_in'."
+            )
+
```

Comment:
Should this validation be done earlier, in the DSL, and raise an error that reports what the user should have written there (in the new-style DSL, that is, when they'd have done something like `events.filter(code==codelist)`?   And/or should the `CodeColumn` in `dsl.py` not define `__eq__` and `__ne__` methods, if we're not allowing those?

Reply:
Oh I think it should definitely be done in the new DSL. This was just a quick fix to catch all the places we were relying on the old behaviour and make sure we don't add any more.

I think the long term goal should be that the DSL does all the user facing validation and can never produce an invalid QM. And then we just have asserts in the QM as a fallback in case we make a mistake in the DSL validation.

Do you think that make sense?